token-metadata/program/src/state.rs

```diff
@@ -75,6 +83,29 @@ pub const USE_AUTHORITY_RECORD_SIZE: usize = 18; //8 byte padding
 
 pub const COLLECTION_AUTHORITY_RECORD_SIZE: usize = 11; //10 byte padding
 
+pub trait TokenMetadataAccount {
+    fn key() -> Key;
+
+    fn size() -> usize;
+
+    fn pad_length(buf: &mut Vec<u8>) {
+        let padding_length = Self::size() - buf.len();
+        buf.extend(vec![0; padding_length]);
+    }
+
+    fn safe_deserialize<T: BorshDeserialize + TokenMetadataAccount>(
```

Comment:
`safe_deserialize` needs to be an associated function because you're trying to get the type from raw bytes, you're not starting with the type. So the typical use case is something like this:

```rust
use mpl_token_metadata::state::{Metadata, TokenMetadataAccount};

. . .
let buffer = client.get_account_data(pubkey)?;
let metadata: Metadata = Metadata::safe_deserialize(&mut buffer)?;
```

I suppose you could create a default or empty `Metadata` type and then you could use it as a method:

```rust
use mpl_token_metadata::state::{Metadata, TokenMetadataAccount};

. . .
let md_empty = Metadata::new();
let buffer = client.get_account_data(pubkey)?;
let metadata = md_empty.safe_deserialize(&mut buffer)?;
```
but this seems awkward and unnecessary.

The same thing goes from `from_account_info` (which I agree should be part of the Trait). Am I missing something?

Reply:
Also, none of these functions actually need any information from `self` so it seems anti-pattern to make them methods.

Reply:
I do think maybe in general it is ok to have a `&self` reference even if you don't know if you need the info from an instance, because when you are just defining a trait you don't restrict the implementation details.  For example:

```
pub trait Summary {
    fn summarize(&self) -> String;
}

struct Something {
    summary: String
}

impl Summary for Something {
    fn summarize(&self) -> String {
        self.summary.clone()
    }
}

struct SomethingElse;

impl Summary for SomethingElse {
    fn summarize(&self) -> String {
        "This summary is hardcoded".to_string()
    }
}
```

I also think `&self` implies you are dealing with an instance and only really want it to be able to be called on an instance.  And I agree with your logic that you are **not** starting with an instance and it doesn't make sense to instantiate one, so I would say you are right.

Also, it looks like you removed the trait bound on the function in the trait (which makes sense to me also).